rfc/merkle-structure.md

```diff
@@ -0,0 +1,492 @@
+
+
+```mermaid
+stateDiagram-v2
+b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq: "hello world"
+[*] --> b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq: b2i…v2tq
+b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq --> bpedmjxitfak5zao5jiabw4ngew6bi3mv4y2fh5li2tdaisujt2ma_b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq  : bpe…t2ma
+bpedmjxitfak5zao5jiabw4ngew6bi3mv4y2fh5li2tdaisujt2ma_b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq: merkle-structureːstring/utf-8
+bytes_b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq: 0x68 0x65 0x6C…0x72 0x6C 0x64
+b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq --> bytes_b2ip5bcmbwyfmckglvjbttorkwz4seqyqpyq425g6iyvyf2d6v2tq
+```
+
+#### Integer
+
+Integer is represented as a [merkle fold] of pair of leaf nodes. Left node is the cryptographic hash of the UTF-8 encoded string `merkle-structure:integer/leb128` and describes binary encoding format of the right node. Right node is [LEB128] encoded integer.
+
+```mermaid
+stateDiagram-v2
+b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q: 1985
+[*] --> b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q: b4o…hx4q
+b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q --> b3txd3gwyvlgf7oqtiijwjatz5rq2vlcznuqmsz6nmhq7skt262yq_b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q  : b3t…62yq
+b3txd3gwyvlgf7oqtiijwjatz5rq2vlcznuqmsz6nmhq7skt262yq_b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q: merkle-structureːinteger/leb128
+bytes_b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q: 0xC1 0x0F
+b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q --> bytes_b4ob7njt6ngtc7723fryqym6uemvyvvfntjwphglwe3ytglbwhx4q
+```
+
+#### Float
+
+Float is represented as a [merkle fold] of pair of leaf nodes. Left node is the cryptographic hash of the UTF-8 encoded string `merkle-structure:float/double-precision` and describes binary encoding format of the right node. Right node is a float encoded in [double-precision floating-point format].
+
+```mermaid
+stateDiagram-v2
+bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa: 18.033
+[*] --> bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa: bmj…27fa
+bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa --> b46btqxqak3bstb7pmf2kkbl2ht6zat2xub5lovb4oavki4uysd2q_bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa  : b46…sd2q
+b46btqxqak3bstb7pmf2kkbl2ht6zat2xub5lovb4oavki4uysd2q_bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa: merkle-structureːfloat/double-precision
+bytes_bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa: 0x9C 0xC4 0x20…0x08 0x32 0x40
+bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa --> bytes_bmjrgvd75uynefn3hljzkl2lg4xqthymoqolc22qwtxl2crew27fa
+```
+
+#### Bytes
+
+Bytes is represented as a [merkle fold] of pair of leaf nodes. Left node is the cryptographic hash of the UTF-8 encoded string `merkle-structure:bytes/raw` and describes binary encoding format of the right node. Right node is a byte array of the bytes.
```

Comment:
🤔 I'm starting to wonder if the right node should be an actual bao tree here

Reply:
I'm convinced now that it should be.

Reply:
I'm convinced now that it should be, that way hash of the raw bytes would come out same as the blake3 hash of that content.

Reply:
Blake3 splits byte array into 1024 byte size chunks (last chunk may be shorter). If there is only one chunk that chunk is a the root.

Reply:
I dug deeper into blake3 & bao and sadly compatibility is probably not going to work out. The problem is that `leaf_hash(chunk, n)` and `node_hash(left, right, is_root)` are not defined in terms of `blake3` hash so we would either have to define same algorithm in the merkle-folding or would instead define them using the hashing function (as per spec today).

Seems to me that we could take advantage of all the benefits that blake3/bao offers with a generic hashing function which seems like a better tradeoff here.

That said it is probably worth evaluating how why chunk index `n` is used and if it would make sense to incorporate that and same for the `is_root` parameter.

Reply:
For the reference bao hash and blake3 hash come out the same

https://github.com/oconnor663/bao/blob/e5f01f8c5d11f653eac861ee1a72f0fae5728460/src/encode.rs#L1206-L1208

Although that is because bao leverages blake3 internals to compute leaf hash

https://github.com/oconnor663/bao/blob/e5f01f8c5d11f653eac861ee1a72f0fae5728460/src/encode.rs#L559-L562

And to compute the node hash

https://github.com/oconnor663/bao/blob/e5f01f8c5d11f653eac861ee1a72f0fae5728460/src/encode.rs#L297-L300